File: c/structure/tree/binary_tree/traversal/morris/InOrderTraversal.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct node {
    int value;
    struct node *left;
    struct node *right;
} TreeNode;

typedef TreeNode *pNode;

typedef struct {
    int *values;
    int length;
    int capacity;
} ArrayNode;

void inOrderTraversal(pNode root, ArrayNode *result) {
    result->values = (int *) malloc(sizeof(int) * 64);
    memset(result->values, 0, sizeof(int) * 64);
    result->length = 0;
    result->capacity = 64;

    pNode current = root;
    while (current) {
        if (!current->left) {
            if (result->length >= result->capacity) {
                result->values = (int *) realloc(result->values, sizeof(int) * (result->capacity + 16));
                memset(result->values + result->capacity, 0, sizeof(int) * 16);
                result->capacity += 16;
            }
            result->values[result->length] = current->value;
            ++result->length;

            current = current->right;
            continue;
        }

        pNode aux = current->left;
        while (aux->right && aux->right != current) {
            aux = aux->right;
        }
        if (!aux->right) {
            aux->right = current;
            current = current->left;
        } else {
            if (result->length >= result->capacity) {
                result->values = (int *) realloc(result->values, sizeof(int) * (result->capacity + 16));
                memset(result->values + result->capacity, 0, sizeof(int) * 16);
                result->capacity += 16;
            }
            result->values[result->length] = current->value;
            ++result->length;

            aux->right = NULL;
            current = current->right;
        }
    }
}

File: java/structure/tree/binary_tree/traversal/morris/InOrderTraversal.java
package structure.tree.binary_tree.traversal.morris;

import java.util.ArrayList;
import java.util.List;

public class InOrderTraversal {
    public int[] traversal(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        TreeNode current = root;
        while (null != current) {
            if (null == current.left) {
                result.add(current.value);
                current = current.right;
                continue;
            }

            TreeNode aux = current.left;
            while (null != aux.right && aux.right != current) {
                aux = aux.right;
            }
            if (null == aux.right) {
                aux.right = current;
                current = current.left;
            } else {
                result.add(current.value);
                aux.right = null;
                current = current.right;
            }
        }
        return result.stream().mapToInt(Integer::valueOf).toArray();
    }

    public class TreeNode {
        public int value = 0;
        public TreeNode left = null;
        public TreeNode right = null;

        public TreeNode(int value) {
            this.value = value;
        }
    }
}
